File: frontend/src/types/teamPlanning.ts
/**
 * Team Planning Types - Phase 5+6
 * 
 * CRITICAL BUSINESS RULES:
 * 1. Status is auto-calculated, never manually set
 * 2. NO locked/is_locked fields (approved items can be modified)
 * 3. Orphaned items preserve JIRA key/title when deleted
 * 4. Notifications have NO expires_at field
 */

export type PlanningStatus = 
  | 'not_planned'      // No role breakdown added
  | 'accepted'         // PO kept PM's effort + added breakdown
  | 'modified'         // PO changed effort from PM's original
  | 'descope_proposed' // PO wants to descope
  | 'orphaned';        // JIRA was deleted while PO was planning

export type ReviewStatus = 'pending' | 'approved' | 'rejected' | null;

export type CapacityStatus = 'green' | 'amber' | 'red' | 'warning';

export type PlanVersionStatus = 'draft' | 'committed' | 'approved' | 'rejected' | 'outdated';

export interface TeamPlanningItem {
  id: string;
  jira_record_id: string | null;  // NULL when orphaned
  jira_key: string;
  jira_title: string;
  feature_name: string;
  team_id: string;
  pi_id: string;
  version_id: string;
  
  // Effort data
  original_pm_effort: number;  // PM's original value
  planned_effort: number | null;
  dev_effort: number;
  pd_effort: number;
  qa_effort: number;
  
  // Status (auto-calculated by backend)
  status: PlanningStatus;
  delta: number | null;  // planned_effort - original_pm_effort
  
  // Descope workflow
  is_descoped: boolean;
  descope_reason: string | null;
  descoped_at: string | null;
  
  // Orphan tracking (JIRA deleted while planning)
  is_orphaned: boolean;
  orphaned_jira_key: string | null;
  orphaned_jira_title: string | null;
  orphaned_at: string | null;
  
  // PM Review (NO locked field)
  review_status: ReviewStatus;
  reviewed_at: string | null;
  reviewed_by: string | null;
  review_note: string | null;
  rejection_reason: string | null;
  
  // Commit tracking
  committed_at: string | null;
  
  // Flags
  is_spillover: boolean;
  
  // Timestamps
  created_at: string;
  updated_at: string;
}

export interface RoleCapacity {
  available: number;
  used: number;
  remaining: number;
}

export interface TeamCapacity {
  available_ed: number;
  used_ed: number;
  remaining_ed: number;
  utilization_percent: number;
  status: CapacityStatus;
  warning?: string;
  roles?: {
    dev: RoleCapacity;
    pd: RoleCapacity;
    qa: RoleCapacity;
  };
}

export interface TeamInfo {
  id: string;
  name: string;
}

export interface PIInfo {
  id: string;
  name: string;
  year: number;
  sequence: number;
}

export interface VersionInfo {
  id: string;
  version_name: string;
  status: string;
}

export interface PlanningSummary {
  total: number;
  accepted: number;
  modified: number;
  descoped: number;
  not_planned: number;
  orphaned: number;
}

export interface TeamPlanningResponse {
  team: TeamInfo;
  pi: PIInfo;
  version: VersionInfo;
  capacity: TeamCapacity;
  items: TeamPlanningItem[];
  summary: PlanningSummary;
  is_outdated?: boolean;
  outdated_reason?: string | null;
  outdated_at?: string | null;
}

export interface POPlanVersion {
  id: string;
  team_id: string;
  pi_id: string;
  strategic_version_id: string;
  version_number: 1 | 2;  // Max 2 versions
  status: PlanVersionStatus;
  committed_at: string | null;
  committed_by: string | null;
  planning_snapshot?: any;  // Preserved data for outdated drafts
  created_at: string;
  updated_at: string;
}

export interface PlanningNotification {
  id: string;
  team_id: string;
  pi_id: string;
  product_id: string;
  notification_type: 'plan_committed' | 'plan_approved' | 'plan_rejected' | 'version_changed' | 'plan_needs_revision';
  message: string;
  team_name: string;
  pi_name: string;
  product_name: string;
  items_count: number;
  total_effort_change: number;
  is_read: boolean;
  read_at: string | null;
  created_at: string;
  // NOTE: NO expires_at field - notifications persist until read
}

// Request/Response types
export interface CreatePlanningRequest {
  jira_record_id: string | null;  // NULL for orphaned items
  team_id: string;
  pi_id: string;
  version_id: string;
  dev_effort: number;
  pd_effort: number;
  qa_effort: number;
}

export interface UpdatePlanningRequest {
  dev_effort?: number;
  pd_effort?: number;
  qa_effort?: number;
}

export interface DescopeRequest {
  reason: string;  // Min 10 chars, max 500
}

export interface CommitPlanRequest {
  pi_id: string;
  version_id?: string;  // Optional - backend will use single draft
}

export interface CommitPlanResponse {
  plan_version_id: string;
  committed_at: string;
  items_count: number;
  notification_sent: boolean;
  summary: PlanningSummary;
}

export interface ApproveItemRequest {
  note?: string;
}

export interface RejectItemRequest {
  reason: string;
}

// Filters
export interface TeamPlanningFilters {
  status?: PlanningStatus[];
  search?: string;
  show_descoped?: boolean;
  show_orphaned?: boolean;
}

File: backend/app/schemas/team_planning.py
"""
Team Planning Schemas - Phase 5A

CRITICAL BUSINESS RULES:
1. Status is auto-calculated, never manually set
2. Capacity thresholds: <95% green, 95-100% amber, >100% red
3. No locked fields
4. No expires_at in notifications
"""
from pydantic import BaseModel, Field, validator
from typing import Optional, List
from datetime import datetime
from decimal import Decimal


class RoleBreakdown(BaseModel):
    """Role breakdown for Dev/PD/QA effort."""
    dev_effort: Decimal = Field(default=0, ge=0, description="Developer effort in eD")
    pd_effort: Decimal = Field(default=0, ge=0, description="Product Designer effort in eD")
    qa_effort: Decimal = Field(default=0, ge=0, description="QA effort in eD")
    
    @validator('dev_effort', 'pd_effort', 'qa_effort')
    def validate_non_negative(cls, v):
        if v < 0:
            raise ValueError('Effort cannot be negative')
        return v


class TeamPlanningBase(BaseModel):
    """Base schema for team planning."""
    jira_record_id: str
    team_id: str
    pi_id: str
    version_id: Optional[str] = None  # Optional - backend gets from JIRA record
    po_plan_version_id: Optional[str] = None  # Which PO draft this belongs to
    planned_effort: Optional[Decimal] = None
    dev_effort: Decimal = Field(default=0, ge=0)
    pd_effort: Decimal = Field(default=0, ge=0)
    qa_effort: Decimal = Field(default=0, ge=0)


class TeamPlanningCreate(TeamPlanningBase):
    """Create team planning record (auto-save)."""
    
    @validator('dev_effort', 'pd_effort', 'qa_effort')
    def validate_non_negative(cls, v):
        if v < 0:
            raise ValueError('Effort cannot be negative')
        return v
    
    @validator('planned_effort', always=True)
    def calculate_planned_effort(cls, v, values):
        """Auto-calculate planned_effort from role breakdown."""
        if 'dev_effort' in values and 'pd_effort' in values and 'qa_effort' in values:
            return values['dev_effort'] + values['pd_effort'] + values['qa_effort']
        return v


class TeamPlanningUpdate(BaseModel):
    """Update team planning record."""
    planned_effort: Optional[Decimal] = None
    dev_effort: Optional[Decimal] = Field(None, ge=0)
    pd_effort: Optional[Decimal] = Field(None, ge=0)
    qa_effort: Optional[Decimal] = Field(None, ge=0)


class TeamPlanningResponse(BaseModel):
    """Team planning response with auto-calculated status."""
    id: str
    jira_record_id: Optional[str]
    team_id: str
    pi_id: str
    version_id: str
    
    # Planning data
    planned_effort: Optional[Decimal]
    dev_effort: Decimal
    pd_effort: Decimal
    qa_effort: Decimal
    
    # Status (auto-calculated)
    status: str  # not_planned, accepted, modified, descope_proposed, orphaned
    original_pm_effort: Optional[Decimal]
    delta: Optional[Decimal] = None
    
    # Descope
    is_descoped: bool
    descope_reason: Optional[str]
    descoped_at: Optional[datetime]
    
    # Orphan
    is_orphaned: bool
    orphaned_jira_key: Optional[str]
    orphaned_jira_title: Optional[str]
    orphaned_at: Optional[datetime]
    
    # Review
    review_status: Optional[str]
    rejection_reason: Optional[str]
    reviewed_at: Optional[datetime]
    
    # Commit
    committed_at: Optional[datetime]
    
    # Joined data (for non-orphaned)
    jira_key: Optional[str] = None
    jira_title: Optional[str] = None
    feature_name: Optional[str] = None
    is_spillover: bool = False
    
    # Audit
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class RoleCapacity(BaseModel):
    """Capacity breakdown by role."""
    available: float
    used: float
    remaining: float


class CapacityResponse(BaseModel):
    """
    Capacity response with EXACT thresholds:
    - < 95% = green
    - 95-100% = amber
    - > 100% = red
    """
    available_ed: float
    used_ed: float
    remaining_ed: float
    utilization_percent: float
    status: str  # green, amber, red, warning
    warning: Optional[str] = None
    roles: Optional[dict] = None  # {"dev": RoleCapacity, "pd": RoleCapacity, "qa": RoleCapacity}


class PlanningSummary(BaseModel):
    """Summary of planning items by status."""
    total: int
    accepted: int
    modified: int
    descoped: int
    not_planned: int
    orphaned: int


class TeamInfo(BaseModel):
    """Team information."""
    id: str
    name: str


class PIInfo(BaseModel):
    """PI information."""
    id: str
    name: str
    year: int
    sequence: int


class VersionInfo(BaseModel):
    """Version information."""
    id: str
    version_name: str
    status: str


class TeamPlanningListResponse(BaseModel):
    """Response for team planning list."""
    team: TeamInfo
    pi: PIInfo
    version: Optional[VersionInfo]
    capacity: CapacityResponse
    items: List[TeamPlanningResponse]
    summary: PlanningSummary
    is_outdated: bool = False
    outdated_reason: Optional[str] = None
    outdated_at: Optional[str] = None


class DescopeRequest(BaseModel):
    """Request to descope an item."""
    reason: str = Field(..., min_length=10, max_length=500, description="Reason for descoping (10-500 chars)")


class RestoreRequest(BaseModel):
    """Request to restore a descoped item."""
    pass


class CommitPlanRequest(BaseModel):
    """Request to commit plan for PM review."""
    pi_id: str
    version_id: Optional[str] = None  # Optional - backend will use single draft


class CommitPlanResponse(BaseModel):
    """Response after committing plan."""
    plan_version_id: str
    committed_at: datetime
    items_count: int
    notification_sent: bool
    summary: PlanningSummary


class AcknowledgeOrphanRequest(BaseModel):
    """Request to acknowledge and remove orphaned item."""
    pass
